File: stardroid-v1/app/src/main/java/com/google/android/stardroid/renderer/util/LabelCollisionResolver.java
package com.google.android.stardroid.renderer.util;

import java.util.ArrayList;
import java.util.List;

/** Resolves overlapping screen-space label bounds by moving later labels downward. */
public final class LabelCollisionResolver {
  private static final int DEFAULT_MAX_ITERATIONS = 10;

  private LabelCollisionResolver() {}

  public static List<LabelPosition> resolveCollisions(List<LabelPosition> labels) {
    return resolveCollisions(labels, DEFAULT_MAX_ITERATIONS);
  }

  public static List<LabelPosition> resolveCollisions(
      List<LabelPosition> labels, int maxIterations) {
    List<LabelPosition> resolvedLabels = new ArrayList<>(labels.size());
    for (LabelPosition label : labels) {
      LabelPosition resolvedLabel = label;
      int iterations = 0;
      while (iterations < maxIterations && overlapsAny(resolvedLabel, resolvedLabels)) {
        resolvedLabel =
            new LabelPosition(
                resolvedLabel.getX(),
                resolvedLabel.getY() - resolvedLabel.getHeight(),
                resolvedLabel.getWidth(),
                resolvedLabel.getHeight());
        iterations++;
      }
      resolvedLabels.add(resolvedLabel);
    }
    return resolvedLabels;
  }

  private static boolean overlapsAny(LabelPosition label, List<LabelPosition> placedLabels) {
    for (LabelPosition placedLabel : placedLabels) {
      if (overlaps(label, placedLabel)) {
        return true;
      }
    }
    return false;
  }

  private static boolean overlaps(LabelPosition first, LabelPosition second) {
    float firstHalfWidth = first.getWidth() * 0.5f;
    float firstHalfHeight = first.getHeight() * 0.5f;
    float secondHalfWidth = second.getWidth() * 0.5f;
    float secondHalfHeight = second.getHeight() * 0.5f;
    return first.getX() - firstHalfWidth < second.getX() + secondHalfWidth
        && first.getX() + firstHalfWidth > second.getX() - secondHalfWidth
        && first.getY() - firstHalfHeight < second.getY() + secondHalfHeight
        && first.getY() + firstHalfHeight > second.getY() - secondHalfHeight;
  }

  /** Screen-space center point and dimensions for a label. */
  public static final class LabelPosition {
    private final float x;
    private final float y;
    private final float width;
    private final float height;

    public LabelPosition(float x, float y, float width, float height) {
      this.x = x;
      this.y = y;
      this.width = width;
      this.height = height;
    }

    public float getX() {
      return x;
    }

    public float getY() {
      return y;
    }

    public float getWidth() {
      return width;
    }

    public float getHeight() {
      return height;
    }
  }
}

File: stardroid-v1/app/src/main/java/com/google/android/stardroid/renderer/util/LabelCollisionResolver.kt
package com.google.android.stardroid.renderer.util

import kotlin.math.cos
import kotlin.math.sin

/** Resolves overlapping label bounds in their shared rotated coordinate system. */
class LabelCollisionResolver {
    private val positions = ArrayList<LabelPosition>()
    private var placedCount = 0

    fun beginFrame() {
        placedCount = 0
    }

    @JvmOverloads
    fun place(
        x: Float,
        y: Float,
        width: Float,
        height: Float,
        upAngle: Float,
        maxIterations: Int = DEFAULT_MAX_ITERATIONS,
    ): LabelPosition {
        val cosine = cos(upAngle)
        val sine = sin(upAngle)
        val localX = cosine * x - sine * y
        var localY = sine * x + cosine * y
        var iterations = 0
        while (
            iterations < maxIterations &&
                overlapsAny(localX, localY, width, height, cosine, sine)
        ) {
            // Keep later labels visible; draw order defines which label retains its anchor.
            localY -= height
            iterations++
        }
        val position =
            positions.getOrNull(placedCount) ?: LabelPosition().also { positions.add(it) }
        position.x = cosine * localX + sine * localY
        position.y = -sine * localX + cosine * localY
        position.width = width
        position.height = height
        placedCount++
        return position
    }

    private fun overlapsAny(
        x: Float,
        y: Float,
        width: Float,
        height: Float,
        cosine: Float,
        sine: Float,
    ): Boolean {
        for (index in 0 until placedCount) {
            val placed = positions[index]
            val placedX = cosine * placed.x - sine * placed.y
            val placedY = sine * placed.x + cosine * placed.y
            if (
                x - width / 2 < placedX + placed.width / 2 &&
                    x + width / 2 > placedX - placed.width / 2 &&
                    y - height / 2 < placedY + placed.height / 2 &&
                    y + height / 2 > placedY - placed.height / 2
            ) {
                return true
            }
        }
        return false
    }

    data class LabelPosition(
        var x: Float = 0f,
        var y: Float = 0f,
        var width: Float = 0f,
        var height: Float = 0f,
    )

    private companion object {
        const val DEFAULT_MAX_ITERATIONS = 10
    }
}
